ejb3/src/main/java/org/jboss/as/ejb3/deployment/processors/DiscoveryService.java

```diff
@@ -22,45 +22,89 @@
     private static final DiscoveryProvider[] NO_PROVIDERS = new DiscoveryProvider[0];
-
     public static final ServiceName BASE_NAME = ServiceName.JBOSS.append("deployment", "discovery");
 
-    private final InjectedSetValue<DiscoveryProvider> providerInjectors = new InjectedSetValue<>();
+    private final Collection<DiscoveryProvider> providers = new LinkedHashSet<>();
+    private final Consumer<Discovery> discoveryConsumer;
+    private final Supplier<RemotingProfileService> remotingProfileServiceSupplier;
+    private final Supplier<AssociationService> associationServiceSupplier;
 
-    private volatile Discovery discovery;
+    private volatile DiscoveryProvider remotingDP;
+    private volatile DiscoveryProvider assocationDP;
 
-    public void start(final StartContext context) throws StartException {
-        discovery = Discovery.create(providerInjectors.getValue().toArray(NO_PROVIDERS));
+    public DiscoveryService(final Consumer<Discovery> discoveryConsumer, final Supplier<RemotingProfileService> remotingProfileServiceSupplier, final Supplier<AssociationService> associationServiceSupplier) {
+        this.discoveryConsumer = discoveryConsumer;
+        this.remotingProfileServiceSupplier = remotingProfileServiceSupplier;
+        this.associationServiceSupplier = associationServiceSupplier;
     }
 
-    public void stop(final StopContext context) {
-        discovery = null;
+    public void start(final StartContext context) throws StartException {
+        final RemotingProfileService remotingProfileService = remotingProfileServiceSupplier != null ? remotingProfileServiceSupplier.get() : null;
+        final AssociationService associationService = associationServiceSupplier != null ? associationServiceSupplier.get() : null;
+
+        if (remotingProfileService != null) {
+            remotingDP = new StaticDiscoveryProvider(remotingProfileService.getServiceUrls());
+        }
+        if (associationService != null) {
+            assocationDP = associationService.getLocalDiscoveryProvider();
+        }
+        final DiscoveryProvider[] providersArray;
+        synchronized (providers) {
+            if (remotingDP != null) providers.add(remotingDP);
+            if (assocationDP != null) providers.add(assocationDP);
+            providersArray = providers.toArray(NO_PROVIDERS);
+        }
+        discoveryConsumer.accept(Discovery.create(providersArray));
     }
 
-    public Injector<DiscoveryProvider> getDiscoveryProviderInjector() {
-        return providerInjectors.injector();
+    public void stop(final StopContext context) {
+        discoveryConsumer.accept(null);
+        synchronized (providers) {
+            if (remotingDP != null) {
+                providers.remove(remotingDP);
+                remotingDP = null;
+            }
+            if (assocationDP != null) {
+                providers.remove(assocationDP);
+                assocationDP = null;
+            }
```

Comment:
do we also need to remove any extra discovery providers added via line 105?

Reply:
Looking at the code I don't think so @chengfang - these are registered via DiscoveryRegistrationProcessor (lines 73 & 74)
DiscoveryService can be restarted any time so everything done in start() method must be undone in stop() method.

Reply:
To be more precise @chengfang this PR doesn't change anything from functional point of view.
It is just one refactoring step that only refactors dependency configuration staff to use a new MSC value API.
Notice that both external methods (in DiscoveryRegistrationProcessor code):
 * HttpDiscoveryConfigurator.configure() 
 * RemoteEJBDiscoveryConfigurator.configure()

are just injecting new provider into DiscoveryService.
This is what was happening before this PR and this what will be happening after this PR (if merged).